contracts/src/contracts/hooks/domain_routing_hook.cairo

```diff
@@ -0,0 +1,108 @@
+#[starknet::contract]
+pub mod domain_routing_hook {
+    use alexandria_bytes::{Bytes, BytesTrait, BytesStore};
+    use hyperlane_starknet::contracts::client::mailboxclient_component::{
+        MailboxclientComponent, MailboxclientComponent::MailboxClientInternalImpl
+    };
+    use hyperlane_starknet::contracts::client::{mailboxclient};
+    use hyperlane_starknet::contracts::libs::message::Message;
+    use hyperlane_starknet::interfaces::{
+        IPostDispatchHook, IPostDispatchHookDispatcher, IPostDispatchHookDispatcherTrait,
+        DomainRoutingHookConfig, IDomainRoutingHook, Types
+    };
+    use openzeppelin::access::ownable::OwnableComponent;
+    use openzeppelin::upgrades::{interface::IUpgradeable, upgradeable::UpgradeableComponent};
+    use starknet::{ContractAddress, contract_address_const};
+    component!(path: OwnableComponent, storage: ownable, event: OwnableEvent);
+    component!(path: MailboxclientComponent, storage: mailboxclient, event: MailboxclientEvent);
+    component!(path: UpgradeableComponent, storage: upgradeable, event: UpgradeableEvent);
+    #[abi(embed_v0)]
+    impl OwnableImpl = OwnableComponent::OwnableImpl<ContractState>;
+    impl OwnableInternalImpl = OwnableComponent::InternalImpl<ContractState>;
+    impl UpgradeableInternalImpl = UpgradeableComponent::InternalImpl<ContractState>;
+
+
+    #[storage]
+    struct Storage {
+        hooks: LegacyMap<u32, IPostDispatchHookDispatcher>,
+        #[substorage(v0)]
+        mailboxclient: MailboxclientComponent::Storage,
+        #[substorage(v0)]
+        ownable: OwnableComponent::Storage,
+        #[substorage(v0)]
+        upgradeable: UpgradeableComponent::Storage,
+    }
+
+
+    mod Errors {
+        pub const INVALID_DESTINATION: felt252 = 'Destination has no hooks';
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        #[flat]
+        OwnableEvent: OwnableComponent::Event,
+        #[flat]
+        UpgradeableEvent: UpgradeableComponent::Event,
+        #[flat]
+        MailboxclientEvent: MailboxclientComponent::Event,
+    }
+
+    #[constructor]
+    fn constructor(ref self: ContractState, _mailbox: ContractAddress, _owner: ContractAddress) {
+        self.mailboxclient.initialize(_mailbox);
+        self.ownable.initializer(_owner);
+    }
+
+    #[abi(embed_v0)]
+    impl IPostDispatchHookImpl of IPostDispatchHook<ContractState> {
+        fn hook_type(self: @ContractState) -> Types {
+            Types::ROUTING(())
+        }
+
+        fn supports_metadata(self: @ContractState, _metadata: Bytes) -> bool {
+            true
+        }
+
+        fn post_dispatch(ref self: ContractState, _metadata: Bytes, _message: Message) {
+            self._get_configured_hook(_message.clone()).post_dispatch(_metadata, _message);
```

Comment:
The post dispatch function lacks the fee mecanism (represented on the solidity version by msg.value). Since we do not have native token, we will have to transfer erc20 manually. Leave it as it is right, now, I will handle it.

Reply:
I can add the fee mechanism. Context would be the `protocol_fee`  hook. Additionally, did you have a look at my test files

Reply:
Yes, protocol fee hook is the context and since it is a routing hook, it will require to quote_dispatch.